File: jhora/horoscope/dhasa/raasi/sudasa.py
from hora import const,utils
from hora.panchanga import drik
from hora.horoscope.chart import house
from hora.horoscope.dhasa.raasi import narayana
from hora.horoscope.chart import charts
def sudasa_dhasa_bhukthi(dob,tob,place,divisional_chart_factor=1,include_antardhasa=True):
    jd = utils.julian_day_number(dob, tob)
    sl = drik.sree_lagna(jd, place, divisional_chart_factor)
    sree_lagna_house = sl[0]
    sree_lagna_longitude = sl[0]*30+sl[1]
    #print('sree_lagna_house',sree_lagna_house,'sree_lagna_longitude',sree_lagna_longitude)
    planet_positions = charts.divisional_chart(jd, place, ayanamsa_mode=const._DEFAULT_AYANAMSA_MODE, divisional_chart_factor=divisional_chart_factor)
    return sudasa_dhasa_from_planet_positions(planet_positions,sree_lagna_house,sree_lagna_longitude,dob,tob,include_antardhasa=include_antardhasa)
def sudasa_dhasa_from_planet_positions(planet_positions,sree_lagna_house,sree_lagna_longitude,dob,tob,include_antardhasa=True):
    """
        calculate Sudasa Dhasa
        @param chart: house_to_planet_list
          Example: ['','','','','2','7','1/5','0','3/4','L','','6/8'] 1st element is Aries and last is Pisces
        @param sree_lagna_house:Raasi index where sree lagna is
        @param sree_lagna_longitude: Longitude of Sree Lagna 
            Note: one can get sree lagna information from drik.sree_lagna()
        @param dob: Date of birth as a tuple e.g. (1999,12,31)  
        @return: 2D list of [dhasa_lord,dhasa_start,[Bhukthi_lord1,bhukthi_lord2,], dhasa_duraation
          Example: [ [7, '1993-6-1', '1996-6-1', [7, 8, 9, 10, 11, 0, 1, 2, 3, 4, 5, 6], 3], ...]
    """
    start_jd = utils.julian_day_number(dob, tob)
    h_to_p = utils.get_house_planet_list_from_planet_positions(planet_positions)
    p_to_h = utils.get_planet_to_house_dict_from_chart(h_to_p)    
    direction = 1
    if sree_lagna_house in const.even_signs:
        direction = -1
    ks = sum(house.kendras()[:3],[])
    """  
        Exceptions: There are exceptions for Saturn or Ketu. If Saturn occupies
                    SL, the counting is forward (no matter odd/even sign). If Ketu occupies
                    SL, the direction of counting houses is reversed from whatever we get based
                    on odd/even sign.
    """
    if p_to_h[6]==sree_lagna_house:
        direction = 1
    elif p_to_h[8]==sree_lagna_house:
        direction *= -1
    dhasa_progression = [(sree_lagna_house+direction*(k-1))%12 for k in ks]
    dhasa_info = []
    for s,dhasa_lord in enumerate(dhasa_progression):
        dhasa_duration = round(narayana._dhasa_duration(planet_positions,dhasa_lord),2)
        if include_antardhasa:
            bhukthis = _antardhasa(dhasa_lord,p_to_h)
            dd = dhasa_duration/12
            for bhukthi_lord in bhukthis:
                y,m,d,h = utils.jd_to_gregorian(start_jd)
                dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
                dhasa_info.append((dhasa_lord,bhukthi_lord,dhasa_start,dd))
                start_jd += dd * const.sidereal_year
        else:
            y,m,d,h = utils.jd_to_gregorian(start_jd)
            dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
            dhasa_info.append((dhasa_lord,dhasa_start,dhasa_duration))
            start_jd += dhasa_duration * const.sidereal_year
    # Second cycle
    dhasa_start = start_jd
    total_dhasa_duration = sum([row[-1] for row in dhasa_info ])
    for c,dhasa_lord in enumerate(dhasa_progression):
        dhasa_duration = 12 - dhasa_info[c][-1]
        dhasa_duration = round(dhasa_duration,2)
        total_dhasa_duration += dhasa_duration
        if dhasa_duration <=0: # no need for second cycle as first cycle had 12 years
            continue
        if include_antardhasa:
            bhukthis = _antardhasa(dhasa_lord,p_to_h)
            dd = dhasa_duration/12
            for bhukthi_lord in bhukthis:
                y,m,d,h = utils.jd_to_gregorian(start_jd)
                dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
                dhasa_info.append((dhasa_lord,bhukthi_lord,dhasa_start,dd))
                start_jd += dd * const.sidereal_year
        else:
            y,m,d,h = utils.jd_to_gregorian(start_jd)
            dhasa_start = '%04d-%02d-%02d' %(y,m,d) +' '+utils.to_dms(h, as_string=True)
            dhasa_info.append((dhasa_lord,dhasa_start,dhasa_duration))
            start_jd += dhasa_duration * const.sidereal_year
        if total_dhasa_duration >= const.human_life_span_for_narayana_dhasa:
            break
    return dhasa_info
def _antardhasa(antardhasa_seed_rasi,p_to_h):
    direction = -1
    if p_to_h[6]==antardhasa_seed_rasi or antardhasa_seed_rasi in const.odd_signs: # Forward
        direction = 1
    if p_to_h[8]==antardhasa_seed_rasi:
        direction *= -1
    return [(antardhasa_seed_rasi+direction*i)%12 for i in range(12)]
if __name__ == "__main__":
    from hora.tests import pvr_tests
    pvr_tests.sudasa_tests()

File: pyjhora-web/src/core/dhasa/raasi/sudasa.ts
/**
 * Sudasa Dasha System
 * Ported from PyJHora sudasa.py
 *
 * Sudasa uses Sree Lagna as the seed. Duration is calculated using
 * Narayana dasha logic. Two cycles are used to cover the life span.
 *
 * Note: Python code mentions that book examples match but JHora differs.
 */

import {
  RASI_NAMES_EN,
  SIDEREAL_YEAR,
  EVEN_SIGNS,
  ODD_SIGNS,
  SATURN,
  KETU
} from '../../constants';
import { PlanetPosition, getDivisionalChart } from '../../horoscope/charts';
import { getHouseOwnerFromPlanetPositions, getPlanetToHouseDict } from '../../horoscope/house';
import { getPlanetLongitude, getSreeLagna } from '../../panchanga/drik';
import { getNarayanaDashaDuration } from './narayana';
import type { Place } from '../../types';
import { julianDayToGregorian } from '../../utils/julian';

// ============================================================================
// TYPES
// ============================================================================

export interface SudasaDashaPeriod {
  rasi: number;
  rasiName: string;
  startJd: number;
  startDate: string;
  durationYears: number;
}

export interface SudasaBhuktiPeriod {
  dashaRasi: number;
  bhuktiRasi: number;
  bhuktiRasiName: string;
  startJd: number;
  startDate: string;
  durationYears: number;
}

export interface SudasaResult {
  mahadashas: SudasaDashaPeriod[];
  bhuktis?: SudasaBhuktiPeriod[];
}

// ============================================================================
// CONSTANTS
// ============================================================================

const YEAR_DURATION = SIDEREAL_YEAR;
const HUMAN_LIFE_SPAN = 120; // Standard paramayush

// Kendra progression: 1,4,7,10,2,5,8,11,3,6,9,12
const KENDRA_PROGRESSION = [1, 4, 7, 10, 2, 5, 8, 11, 3, 6, 9, 12];

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

function getPlanetPositionsArray(jd: number, place: Place, divisionalChartFactor: number): PlanetPosition[] {
  const d1Positions: PlanetPosition[] = [];

  for (let planet = 0; planet <= 8; planet++) {
    const longitude = getPlanetLongitude(jd, place, planet);
    d1Positions.push({
      planet,
      rasi: Math.floor(longitude / 30),
      longitude: longitude % 30
    });
  }

  if (divisionalChartFactor > 1) {
    return getDivisionalChart(d1Positions, divisionalChartFactor);
  }

  return d1Positions;
}

function formatJdAsDate(jd: number): string {
  const { date, time } = julianDayToGregorian(jd);
  const pad = (n: number) => Math.abs(n).toString().padStart(2, '0');
  const hour12 = time.hour % 12 || 12;
  const ampm = time.hour < 12 ? 'AM' : 'PM';
  const yearStr = date.year < 0 ? `${Math.abs(date.year)} BC` : date.year.toString();
  return `${yearStr}-${pad(date.month)}-${pad(date.day)} ${pad(hour12)}:${pad(time.minute)}:${pad(time.second)} ${ampm}`;
}

/**
 * Get antardhasa progression based on seed rasi and planet positions
 */
function getAntardhasa(
  antardhasaSeedRasi: number,
  planetPositions: PlanetPosition[]
): number[] {
  const pToH = getPlanetToHouseDict(planetPositions);

  let direction = -1;
  // Forward if Saturn is in seed or if seed is odd sign
  if (pToH[SATURN] === antardhasaSeedRasi || ODD_SIGNS.includes(antardhasaSeedRasi)) {
    direction = 1;
  }
  // Reverse if Ketu is in seed
  if (pToH[KETU] === antardhasaSeedRasi) {
    direction *= -1;
  }

  return Array.from({ length: 12 }, (_, i) =>
    ((antardhasaSeedRasi + direction * i) % 12 + 12) % 12
  );
}

// ============================================================================
// MAIN FUNCTION
// ============================================================================

/**
 * Get Sudasa Dasha periods
 * Uses Sree Lagna as seed with Narayana-style duration calculation
 *
 * @param jd - Julian day number
 * @param place - Birth place
 * @param options - Configuration options
 * @param options.divisionalChartFactor - Divisional chart factor (default 1 for D-1)
 * @param options.includeBhuktis - Whether to include sub-periods (default true)
 */
export function getSudasaDashaBhukti(
  jd: number,
  place: Place,
  options: {
    divisionalChartFactor?: number;
    includeBhuktis?: boolean;
  } = {}
): SudasaResult {
  const {
    divisionalChartFactor = 1,
    includeBhuktis = true
  } = options;

  const planetPositions = getPlanetPositionsArray(jd, place, divisionalChartFactor);
  const pToH = getPlanetToHouseDict(planetPositions);

  // Get Sree Lagna
  const [sreeLagnaHouse, sreeLagnaLongitude] = getSreeLagna(jd, place);

  // Fraction remaining at birth for first dasha
  const slFracLeft = (30 - sreeLagnaLongitude) / 30;

  // Determine direction based on even/odd sign
  let direction = 1;
  if (EVEN_SIGNS.includes(sreeLagnaHouse)) {
    direction = -1;
  }

  // Saturn/Ketu exceptions
  if (pToH[SATURN] === sreeLagnaHouse) {
    direction = 1;
  } else if (pToH[KETU] === sreeLagnaHouse) {
    direction *= -1;
  }

  // Build progression using kendra pattern
  const dhasaProgression = KENDRA_PROGRESSION.map(k =>
    ((sreeLagnaHouse + direction * (k - 1)) % 12 + 12) % 12
  );

  const mahadashas: SudasaDashaPeriod[] = [];
  const bhuktis: SudasaBhuktiPeriod[] = [];
  let startJd = jd;
  let totalDuration = 0;
  const firstCycleDurations: number[] = [];

  // First cycle
  for (let s = 0; s < dhasaProgression.length; s++) {
    const dhasaLord = dhasaProgression[s]!;
    let duration = getNarayanaDashaDuration(planetPositions, dhasaLord);

    // First dasha uses fraction remaining
    if (s === 0) {
      duration *= slFracLeft;
    }

    duration = Math.round(duration * 100) / 100;
    firstCycleDurations.push(duration);

    const rasiName = RASI_NAMES_EN[dhasaLord] ?? `Rasi ${dhasaLord}`;

    mahadashas.push({
      rasi: dhasaLord,
      rasiName,
      startJd,
      startDate: formatJdAsDate(startJd),
      durationYears: duration
    });

    if (includeBhuktis) {
      const bhuktiLords = getAntardhasa(dhasaLord, planetPositions);
      const bhuktiDuration = duration / 12;
      let bhuktiStartJd = startJd;

      for (const bhuktiLord of bhuktiLords) {
        const bhuktiRasiName = RASI_NAMES_EN[bhuktiLord] ?? `Rasi ${bhuktiLord}`;

        bhuktis.push({
          dashaRasi: dhasaLord,
          bhuktiRasi: bhuktiLord,
          bhuktiRasiName,
          startJd: bhuktiStartJd,
          startDate: formatJdAsDate(bhuktiStartJd),
          durationYears: bhuktiDuration
        });

        bhuktiStartJd += bhuktiDuration * YEAR_DURATION;
      }
    }

    startJd += duration * YEAR_DURATION;
    totalDuration += duration;
  }

  // Second cycle (complement to 12 years each)
  for (let c = 0; c < dhasaProgression.length; c++) {
    if (totalDuration >= HUMAN_LIFE_SPAN) break;

    const dhasaLord = dhasaProgression[c]!;
    const firstDuration = c === 0
      ? getNarayanaDashaDuration(planetPositions, dhasaLord)
      : firstCycleDurations[c]!;

    let duration = 12 - firstDuration;
    duration = Math.round(duration * 100) / 100;

    if (duration <= 0) continue;

    totalDuration += duration;

    const rasiName = RASI_NAMES_EN[dhasaLord] ?? `Rasi ${dhasaLord}`;

    mahadashas.push({
      rasi: dhasaLord,
      rasiName,
      startJd,
      startDate: formatJdAsDate(startJd),
      durationYears: duration
    });

    if (includeBhuktis) {
      const bhuktiLords = getAntardhasa(dhasaLord, planetPositions);
      const bhuktiDuration = duration / 12;
      let bhuktiStartJd = startJd;

      for (const bhuktiLord of bhuktiLords) {
        const bhuktiRasiName = RASI_NAMES_EN[bhuktiLord] ?? `Rasi ${bhuktiLord}`;

        bhuktis.push({
          dashaRasi: dhasaLord,
          bhuktiRasi: bhuktiLord,
          bhuktiRasiName,
          startJd: bhuktiStartJd,
          startDate: formatJdAsDate(bhuktiStartJd),
          durationYears: bhuktiDuration
        });

        bhuktiStartJd += bhuktiDuration * YEAR_DURATION;
      }
    }

    startJd += duration * YEAR_DURATION;
  }

  return includeBhuktis ? { mahadashas, bhuktis } : { mahadashas };
}
